### PR TITLE
Fix PVS Studio errors

### PR DESCRIPTION
**Pull request by amankumar1**
Investigate and fix the first batch of PVS studio errors in the codebase. See #413.

### CODE REVIEW

File: compiler/ras/DebugCounter.cpp [hunk 1]
```diff
@@ -304,10 +304,13 @@ void TR::DebugCounterAggregation::aggregateStandardCounters(TR::Compilation *com
 void TR::DebugCounterAggregation::aggregateDebugCounterInsertions(TR::Compilation *comp, TR::Node *node, TR::DebugCounter *counter, int32_t delta, int8_t fidelity, int32_t staticDelta)
    {
    const char *counterNames[3];
-   counter->getInsertionCounterNames(comp, node, counterNames);
 
-   if (counter && counter->getDenominator() && counter->contributesToDenominator())
-      aggregateDebugCounterInsertions(comp, node, counter->getDenominator(), delta, fidelity, staticDelta);
+   if (counter)
```
Comment (andrewcraik) [inline compiler/ras/DebugCounter.cpp]: This needs an else IMO - the insidious case of counter being invalid is a silent failure by the look of things - not the ideal behaviour.

File: compiler/optimizer/SinkStores.cpp [hunk 2]
```diff
@@ -3415,7 +3415,10 @@ bool TR_TrivialSinkStores::sinkStorePlacement(TR_MovableStore *store,
          _symbolsKilledInBlock[currentBlockNumber]->print(comp());
          traceMsg(comp(),"\n");
          traceMsg(comp(),"         blockNumber %d symbols used:  ",currentBlockNumber);
-         _symbolsUsedInBlock[currentBlockNumber]->print(comp());
+         if (_symbolsUsedInBlock[currentBlockNumber])
```
Comment (andrewcraik) [inline compiler/optimizer/SinkStores.cpp]: Failure to print the diagnostic because of this if should still print something so this needs an else that prints in some way consistent with the trace formatting around it.

File: compiler/optimizer/VPConstraint.cpp [hunk 3]
```diff
@@ -4802,8 +4802,8 @@ bool TR::VPClass::mustBeEqual(TR::VPConstraint *other, TR::ValuePropagation *vp)
       return getKnownObject()->getIndex() == other->getKnownObject()->getIndex();
 
    TR::VPClass *otherClass = NULL;
-   if (other)
-      otherClass = other->asClass();
+   otherClass = other->asClass();
```
Comment (andrewcraik) [inline compiler/optimizer/VPConstraint.cpp]: this should be merged with the otherClass declaration - why declare and then set rather than set when declaring?

File: compiler/optimizer/VPConstraint.cpp [hunk 4]
```diff
@@ -4984,9 +4984,9 @@ bool TR::VPClass::mustBeNotEqual(TR::VPConstraint *other, TR::ValuePropagation *
       return getKnownObject()->getIndex() != other->getKnownObject()->getIndex();
 
    TR::VPClass *otherClass = NULL;
-   if (other)
-      otherClass = other->asClass();
-  if (!_preexistence &&
+   otherClass = other->asClass();
+
+   if (!_preexistence &&
```
Comment (andrewcraik) [inline compiler/optimizer/VPConstraint.cpp]: Can you explain this change - I don't see it based on the code around it...

File: compiler/x/codegen/CompareAnalyser.cpp [hunk 5]
```diff
@@ -730,8 +730,7 @@ void TR_X86CompareAnalyser::longOrderedCompareAndBranchAnalyser(TR::Node       *
    cg()->decReferenceCount(firstChild);
    cg()->decReferenceCount(secondChild);
 
-   if (deps)
-      deps->setMayNeedToPopFPRegisters(true);
+   deps->setMayNeedToPopFPRegisters(true);
```
Comment (andrewcraik) [inline compiler/x/codegen/CompareAnalyser.cpp]: This isn't the only occurrance of this in the method that I can see so we should either fix them all or leave it along IMO.

File: compiler/optimizer/OMROptimizer.cpp [hunk 6]
```diff
@@ -1566,7 +1566,6 @@ int32_t OMR::Optimizer::performOptimization(const OptimizationStrategy *optimiza
             bool blocksArePending = false;
             while (subGroup->_num != endGroup && subGroup->_num != endOpts)
                {
-               OMR::Optimizations optNum = subGroup->_num;
```
Comment (andrewcraik) [inline compiler/optimizer/OMROptimizer.cpp]: I'm not sure this is right the definition of optNum way back in the file is for the optimization overall and not the number within the subgroup - @vijaysun-omr do you agree?

Comment (vijaysun-omr) [inline compiler/optimizer/OMROptimizer.cpp]: Unless I am mistaken, this definition of optNum is "dead" in that no one seems to use this value inside the scope/loop containing the definition.

**Comment by vijaysun-omr** [comment]
One suggestion to consider for this change in particular (since in touches so many files doing quite unrelated things) is splitting these changes into separate PRs based on some logical classification so that each of the PRs can be tested separately. This could help in narrowing down as well as assigning any functional regressions to an owner who would have some expertise in the problematic area.

Just a suggestion to consider.

**Comment by Leonardo2718** [comment]
I don't know if it's worth redoing this work, but I definitely agree with @vijaysun-omr. This should definitely be an approach to consider in future refactoring work.

File: compiler/optimizer/OMROptimizer.cpp (hunk 6 above)
Comment (andrewcraik) [inline compiler/optimizer/OMROptimizer.cpp]: ok if it is dead that is fine - I was more looking at the meaning of the local itself - if unused that should be fine to delete.

**Comment by mgaudet** [comment]
This will also need review from someone who knows the port library @youngar, any suggestions?

File: compiler/optimizer/VPConstraint.cpp (hunk 3 above)
Comment (amankumar1) [inline compiler/optimizer/VPConstraint.cpp]: Didn't notice that it was just declared. Fixed.

File: compiler/optimizer/VPConstraint.cpp [hunk 7]
```diff
@@ -4984,9 +4984,9 @@ bool TR::VPClass::mustBeNotEqual(TR::VPConstraint *other, TR::ValuePropagation *
       return getKnownObject()->getIndex() != other->getKnownObject()->getIndex();
 
    TR::VPClass *otherClass = NULL;
-   if (other)
-      otherClass = other->asClass();
-  if (!_preexistence &&
+   otherClass = other->asClass();
```
Comment (amankumar1) [inline compiler/optimizer/VPConstraint.cpp]: Didn't notice that it was just declared. Fixed.

File: compiler/optimizer/VPConstraint.cpp (hunk 4 above)
Comment (amankumar1) [inline compiler/optimizer/VPConstraint.cpp]: This change is just adding a space before the `if` to fix indentation. I think git got confused with the change right above it.

File: compiler/x/codegen/CompareAnalyser.cpp (hunk 5 above)
Comment (amankumar1) [inline compiler/x/codegen/CompareAnalyser.cpp]: It is checked against nullptr in some cases and is just directly used in other cases. I removed this check because `deps` was just used (see line 717).

File: compiler/optimizer/SinkStores.cpp (hunk 2 above)
Comment (amankumar1) [inline compiler/optimizer/SinkStores.cpp]: Do you have a suggestion for what message to print in the else block?

**Comment by amankumar1** [comment]
Thank you for the suggestions. I have incorporated some of the changes and asked for more input on the other ones.

File: compiler/x/codegen/CompareAnalyser.cpp (hunk 5 above)
Comment (andrewcraik) [inline compiler/x/codegen/CompareAnalyser.cpp]: I don't doubt this one is ok - it just seemed odd to fix it in only one case in the method, but if that is what you prefer to do I'm ok with it. I was just pointing out there may be more that can be tidied up.

File: compiler/ras/DebugCounter.cpp (hunk 1 above)
Comment (andrewcraik) [inline compiler/ras/DebugCounter.cpp]: Having a null counter makes no sense so a TR_ASSERT followed by an early return would make sense to me.

Comment (amankumar1) [inline compiler/ras/DebugCounter.cpp]: Added else block.

**Comment by amankumar1** [comment]
All changed incorporated.

File: compiler/ilgen/IlInjector.cpp [hunk 8]
```diff
@@ -309,11 +309,11 @@ OMR::IlInjector::shiftLeftBy(TR::Node *value, int32_t shift)
    {
    TR::Node *result;
    if (value->getDataType() == TR::Int32)
-      TR::Node::create(TR::ishl, 2, value, TR::Node::iconst(shift));
+      result = TR::Node::create(TR::ishl, 2, value, TR::Node::iconst(shift));
```
Comment (mstoodle) [inline compiler/ilgen/IlInjector.cpp]: Yikes! thanks for fixing that!

File: compiler/optimizer/InductionVariable.cpp [hunk 9]
```diff
@@ -689,7 +689,7 @@ int32_t TR_LoopStrider::detectCanonicalizedPredictableLoops(TR_Structure *loopSt
                   }
 
                ListIterator<TR::Block> bi(&exitBlocks);
-               vcount_t visitCount = comp()->incVisitCount();
+               visitCount = comp()->incVisitCount();
```
Comment (mstoodle) [inline compiler/optimizer/InductionVariable.cpp]: what a monster function. Isn't this the actual definition of the visitCount that's otherwise used in the function? I see the def is higher up and sets to zero but this spot looks like it dominates to me.

File: port/unix/omrsysinfo.c [hunk 10]
```diff
@@ -2201,6 +2199,8 @@ copyEnvToBuffer(struct OMRPortLibrary *portLibrary, void *args)
 		return storageRequiredForEnvironment;
 	}
 
+	memset(buffer, 0, bufferSize);
```
Comment (mstoodle) [inline port/unix/omrsysinfo.c]: that seems like a very good fix!!

File: thread/common/omrthreadinspect.c [hunk 11]
```diff
@@ -296,14 +296,15 @@ omrthread_get_stack_range(omrthread_t thread, void **stackStart, void **stackEnd
 
 #if defined(LINUX)
 	pthread_attr_t attr;
```
Comment (mstoodle) [inline thread/common/omrthreadinspect.c]: Any reason you moved rc and stackSize but not attr?

File: compiler/optimizer/InductionVariable.cpp (hunk 9 above)
Comment (andrewcraik) [inline compiler/optimizer/InductionVariable.cpp]: I don't think it is - I think in this case this was previously local and only and so the dominant definition was either 0 or set in an if higher up the function. the thing that makes this safe is that the local visitcount is not actually used directly. All the helpers called always ask comp()->getVisitCount() and so the key thing here is the inc. The local value is not used downstream from what I can see.

File: tools/tracegen/DATFileWriter.cpp [hunk 12]
```diff
@@ -63,55 +63,53 @@ DATFileWriter::writeOutputFiles(J9TDFOptions *options, J9TDFFile *tdf)
...
-			}
-
-			fprintf(datFile, datLineTemplate, tdf->header.executable, id, tp->type, tp->overhead, tp->level, explicitChar, tp->name, exceptChar, entryExitChar, format);
-			tp = tp->nexttp;
-			id += 1;
-		}
-		fclose(datFile);
-		rc = RC_OK;
-	}
+   while (NULL != tp) {
```
Comment (mstoodle) [inline tools/tracegen/DATFileWriter.cpp]: why did you change the formatting here? you've broken the consistency of the coding standard in this file...

File: compiler/optimizer/GlobalRegisterAllocator.cpp [hunk 13]
```diff
@@ -800,29 +800,35 @@ TR_GlobalRegisterAllocator::restoreOriginalSymbol(TR::Node *node, vcount_t visit
             {
             TR_RegisterCandidate *origRc = _registerCandidates[origSymRef->getReferenceNumber()];
 
-            if (setValueModified)
-               _valueModifiedSymRefs->set(origRc->getSymbolReference()->getReferenceNumber());
+            if (origRc)
```
Comment (mstoodle) [inline compiler/optimizer/GlobalRegisterAllocator.cpp]: Something doesn't seem right here. You've got a conditional structure:
 if (a || b) { if (c) { if (!a && d) ... 

Once you're under (a || b) isn't !b the same as "a" ?

But that makes me suspicious of the "if (!origRc->getValueModified() && origRc->getRestoreSymbolReference()" which could be rewritten as "if (origRc->extendedLiveRange() && origRc->getRestoreSymbolReference())". That feels like a strange condition here, but I didn't look at the whole context.

@andrewcraik or @vijaysun-omr, would you please take a closer look at this code to make sure there isn't a latent bug in the conditions that might be tweaked by this refactoring?

File: compiler/optimizer/LocalOpts.cpp [hunk 14]
```diff
@@ -1166,6 +1166,7 @@ int32_t TR_BlockManipulator::performChecksAndTreesMovement(TR::Block *newBlock,
          ;
 
       TR::Block *nextBlock = prevBlock->getNextBlock();
+      // If this condition is satisfied, nextBlock != NULL
```
Comment (mstoodle) [inline compiler/optimizer/LocalOpts.cpp]: what's the point of adding this comment? And how did you come to this conclusion?

File: compiler/optimizer/LoopReducer.cpp [hunk 15]
```diff
@@ -2761,9 +2761,17 @@ TR_LoopReducer::generateArraytranslate(TR_RegionStructure * whileLoop, TR_Induct
...
             }
          storeNode = (storeTree->getNode()->getNumChildren() == 2) ? storeTree->getNode() : NULL;
+
+         if (storeNode == NULL)
+            {
+            dumpOptDetails(comp(), "Store node is %p\n", storeNode);
+            return false;
+            }
+
          loadTree  = (storeNode != NULL && storeNode->getNumChildren() == 2) ? storeTree : NULL;
```
Comment (mstoodle) [inline compiler/optimizer/LoopReducer.cpp]: wasn't the point of adding the if(storeNode == NULL) block above so that you wouldn't have to test if storeNode is NULL Here?

File: compiler/optimizer/VPConstraint.cpp [hunk 16]
```diff
@@ -4801,9 +4801,8 @@ bool TR::VPClass::mustBeEqual(TR::VPConstraint *other, TR::ValuePropagation *vp)
    if (getKnownObject() && other->getKnownObject() && isNonNullObject() && other->isNonNullObject())
       return getKnownObject()->getIndex() == other->getKnownObject()->getIndex();
 
-   TR::VPClass *otherClass = NULL;
-   if (other)
-      otherClass = other->asClass();
+   TR::VPClass *otherClass = other->asClass();
```
Comment (mstoodle) [inline compiler/optimizer/VPConstraint.cpp]: how did you conclude that other is non-NULL Here?

File: compiler/optimizer/VPConstraint.cpp [hunk 17]
```diff
@@ -4983,10 +4982,9 @@ bool TR::VPClass::mustBeNotEqual(TR::VPConstraint *other, TR::ValuePropagation *
    if (getKnownObject() && other->getKnownObject() && isNonNullObject() && other->isNonNullObject())
       return getKnownObject()->getIndex() != other->getKnownObject()->getIndex();
 
-   TR::VPClass *otherClass = NULL;
-   if (other)
-      otherClass = other->asClass();
-  if (!_preexistence &&
+   TR::VPClass *otherClass = other->asClass();
```
Comment (mstoodle) [inline compiler/optimizer/VPConstraint.cpp]: and here?

**Comment by mstoodle** [comment]
General comment: removing redundant null tests is one of the more dangerous changes you can make (in terms of risk to cause a bug) with a relatively lesser benefit (sure, it cleans up the source code a bit). Several of those changes you made were not immediately obviously correct, and some looked dangerous.

On top of that, this was a large and "painful" PR to review due to the sheer number of completely unrelated changes with potentially dangerous consequences if something is missed.

In future, I would request that you feed these kinds of changes to us in smaller chunks.

File: compiler/optimizer/InductionVariable.cpp (hunk 9 above)
Comment (mstoodle) [inline compiler/optimizer/InductionVariable.cpp]: yeah, that's my bad. i was looking at visitCount's in the file and somehow got myself onto the one at line 4310 which wasn't touched in this commit but looks very similar. Thanks, @andrewcraik .

File: omr/startup/omrvmstartup.cpp [hunk 18]
```diff
@@ -463,6 +463,7 @@ OMR_Shutdown_VM(OMR_VM *omrVM, OMR_VMThread *omrVMThread)
 {
 	omr_error_t rc = OMR_ERROR_NONE;
 
+   // This path assumes that omrVM has been initialized
```
Comment (mstoodle) [inline omr/startup/omrvmstartup.cpp]: Another ping @charliegracie :)

File: omrtrace/omrtracecomponent.cpp [hunk 19]
```diff
@@ -222,15 +222,16 @@ addComponentToList(UtComponentData *componentData, UtComponentList *componentLis
...
-	UT_DBGOUT(1, ("<UT> addComponentToList: component: %s list: %p\n", componentData->componentName, componentList));
-	if (componentList == NULL) {
-		UT_DBGOUT(1, ("<UT> Not adding %s to NULL component list\n", componentData->componentName));
-		return OMR_ERROR_ILLEGAL_ARGUMENT;
-	}
 	if (componentData == NULL) {
 		UT_DBGOUT(1, ("<UT> Not adding NULL component to component list\n"));
 		return OMR_ERROR_ILLEGAL_ARGUMENT;
 	}
+	if (componentList == NULL) {
```
Comment (mstoodle) [inline omrtrace/omrtracecomponent.cpp]: Another ping @charliegracie :)

File: compiler/optimizer/GlobalRegisterAllocator.cpp (hunk 13 above)
Comment (mstoodle) [inline compiler/optimizer/GlobalRegisterAllocator.cpp]: Another ping @andrewcraik , @vijaysun-omr :)

**Comment by mstoodle** [comment]
@amankumar1 lots of good stuff here, but still a few questions requesting clarity before it can be merged...

File: thread/common/omrthreadinspect.c (hunk 11 above)
Comment (amankumar1) [inline thread/common/omrthreadinspect.c]: I moved `attr` too now :)

File: tools/tracegen/DATFileWriter.cpp (hunk 12 above)
Comment (amankumar1) [inline tools/tracegen/DATFileWriter.cpp]: Sorry, I didn't realise this file was using tabs instead of spaces. Fixed.

File: compiler/optimizer/LocalOpts.cpp (hunk 14 above)
Comment (amankumar1) [inline compiler/optimizer/LocalOpts.cpp]: I added this because `nextBlock` is deferenced inside this if statement. I assumed that because of this, `nextBlock` must not be NULL at this point in the code (the warning was caused because of limitations of the static analysis tool). I added this comment for debugging purposes so someone doesn't waste time debugging an issue related to this later. I can delete it if you want.

File: compiler/optimizer/LoopReducer.cpp (hunk 15 above)
Comment (amankumar1) [inline compiler/optimizer/LoopReducer.cpp]: You're right. This was an oversight on my part. Fixed now :)

File: compiler/optimizer/OMRSimplifierHandlers.cpp [hunk 20]
```diff
@@ -3625,8 +3625,7 @@ static bool isLegalToMerge(TR::Node * node, TR::Block * block, TR::Block * nextB
       }
 
    // Do not remove loop pre-header unless it is last run
-   if (cfg &&
-       /* cfg->getStructure() && */
+   if (/* cfg->getStructure() && */
```
Comment (amankumar1) [inline compiler/optimizer/OMRSimplifierHandlers.cpp]: `cfg` could be NULL here. I have added the check back.

File: compiler/optimizer/PartialRedundancy.cpp [hunk 21]
```diff
@@ -2072,8 +2072,7 @@ bool TR_PartialRedundancy::eliminateRedundantSupportedNodes(TR::Node *parent, TR
                   }
                else if (fe()->dataTypeForLoadOrStore(node->getDataType()) != node->getDataType())
                   {
-                  if (parent &&
-                      parent->getOpCode().isConversion() &&
+                  if (parent->getOpCode().isConversion() &&
```
Comment (amankumar1) [inline compiler/optimizer/PartialRedundancy.cpp]: `parent` could be NULL here. I have added the check back.

File: compiler/optimizer/VPConstraint.cpp (hunk 16 above)
Comment (amankumar1) [inline compiler/optimizer/VPConstraint.cpp]: `other` could be NULL here. I have added the check back.

File: compiler/optimizer/VPConstraint.cpp (hunk 17 above)
Comment (amankumar1) [inline compiler/optimizer/VPConstraint.cpp]: `other` could be NULL here. I have added the check back.

**Comment by amankumar1** [comment]
@mstoodle Thank you for reviewing the PR in detail. I have addressed most of the issues you mentioned. Just the comments in which you @mentioned other people to review are pending.

I will separate large chunks into separate PRs in the future :+1:

File: omrtrace/omrtracecomponent.cpp (hunk 19 above)
Comment (charliegracie) [inline omrtrace/omrtracecomponent.cpp]: This looks good to me.  Re-ordering those checks and the debug output is fine.

File: omr/startup/omrvmstartup.cpp (hunk 18 above)
Comment (charliegracie) [inline omr/startup/omrvmstartup.cpp]: I would move the check for omrVM up and wrap everything except the port library and thread library shutdown.

File: compiler/optimizer/GlobalRegisterAllocator.cpp (hunk 13 above)
Comment (andrewcraik) [inline compiler/optimizer/GlobalRegisterAllocator.cpp]: I think I'd rather pull this out of the current pull request - it seems a lot more complicated and worthy of separate consideration.

File: compiler/optimizer/LocalOpts.cpp (hunk 14 above)
Comment (mstoodle) [inline compiler/optimizer/LocalOpts.cpp]: I think the comment is misleading, if it's forecasting that nextBlock is non-NULL just because it is dereferenced inside the next block of code without first checking for NULL. My natural inclination is to suspect a missing NULL test rather than an undocumented non-NULL guarantee. Call me paranoid.

In this particular case, there's a condition below this code block that strongly suggests nextBlock is not necessarily non-NULL in this function. I think the condition that dereferences nextBlock to get its entry tree should actually test nextBlock != NULL first. In other words, I think you found a missing NULL test, and it should be fixed by adding the NULL test, not by adding a comment that implies the condition controlling the unguarded dereference being true means it should not be NULL.

It's still entirely possible I'm missing some subtlety here, but the NULL test doesn't hurt much if the comment you wrote is correct, whereas I think the comment is confusing if nextBlock can, in fact, be NULL (and you'd be leaving a bug in place).

**Comment by fjeremic** [comment]
Is this commit beyond resurrection?

**Comment by andrewcraik** [comment]
I think there were a number of concerns with this change - the very large number of changes in one pull request, the fact the pull request touched code in multiple components without a coupling between those changes being two of the big ones. There were some good clean-ups in here, but there were also some very high-risk logic changes that did not seem obviously correct on inspection. Given the time that has passed it probably makes more sense for this to be broken up and the pieces that make sense delivered piece by piece rather than as one, hard-to-review blob.

**Comment by mgaudet** [comment]
I would agree with Andrew. We need someone to pull this into a large number of individual commits, and then deliver them tiered on risk. We can probably close this, and we'll keep it in mind for something for new people to tackle.

**Comment by mstoodle** [comment]
Closing based on @mgaudet's comment.
